doc/developer/design/2024_10_08_expression_cache.md

```diff
@@ -0,0 +1,135 @@
+# Expression Cache
+
+## The Problem
+
+Optimization is a slow process that is on the critical path for `environmentd` startup times. Some
+environments spend 30 seconds just on optimization. All optimization time spent in startup is
+experienced  downtime for users when `environmentd` restarts.
+
+## Success Criteria
+
+Startup spends less than 1 second optimizing expressions.
+
+## Solution Proposal
+
+The solution being proposed in this document is a cache of optimized expressions. During startup,
+`environmentd` will first look in the cache for optimized expressions and only compute a new
+expression if it isn't present in the cache. If enough expressions are cached and the cache is fast
+enough, then the time spent on this part of startup should be small.
+
+The cache will present similarly as a key-value value store where the key is a composite of
+
+  - deployment generation
+  - object global ID
+  - expression type (local MIR, global MIR, LIR, etc)
```

Comment:
Keep in mind we don't have a serialization format for MIR, and the one for LIR is not designed to be cashed.

Reply:
Yes, this seems fine. (Just be prepared that this will be a lot of typing.)

Reply:
> Just be prepared that this will be a lot of typing

Implementing a serialization format will be a lot of typing?